cms/djangoapps/contentstore/tasks.py

```diff
@@ -1066,3 +1070,186 @@ def undo_all_library_source_blocks_ids_for_course(course_key_string, v1_to_v2_li
             store.update_item(draft_library_source_block, None)
     # return success
     return
+
+
+class CourseLinkCheckTask(UserTask):  # pylint: disable=abstract-method
+    """
+    Base class for course link check tasks.
+    """
+
+    @staticmethod
+    def calculate_total_steps(arguments_dict):
+        """
+        Get the number of in-progress steps in the link check process, as shown in the UI.
+
+        For reference, these are:
+        1. Scanning
+        2. Verifying
+        """
+        return 2
+
+    @classmethod
+    def generate_name(cls, arguments_dict):
+        """
+        Create a name for this particular task instance.
+
+        Arguments:
+            arguments_dict (dict): The arguments given to the task function
+
+        Returns:
+            str: The generated name
+        """
+        key = arguments_dict['course_key_string']
+        return f'Broken link check of {key}'
+
+
+@shared_task(base=CourseLinkCheckTask, bind=True)
+def check_broken_links(self, user_id, course_key_string, language):
+    """
+    Checks for broken links in a course. Store the results in a file.
+    """
+    courselike_key = CourseKey.from_string(course_key_string)
+
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist:
+        with translation_language(language):
+            self.status.fail(UserErrors.UNKNOWN_USER_ID.format(user_id))
+        return
+    if not has_course_author_access(user, courselike_key):
+        with translation_language(language):
+            self.status.fail(UserErrors.PERMISSION_DENIED)
+        return
+    
+    courselike_block = modulestore().get_course(courselike_key)
+
+    try:
+        self.status.set_state('Scanning')
+        links_file = _create_broken_link_json(courselike_block, courselike_key, {}, self.status)
+        artifact = UserTaskArtifact(status=self.status, name='BrokenLinks')
+        artifact.file.save(name=os.path.basename(links_file.name), content=File(links_file))
+        artifact.save()
+    # catch all exceptions so we can record useful error messages
+    except Exception as exception:  # pylint: disable=broad-except
+        LOGGER.exception('Error checking links for course %s', courselike_key, exc_info=True)
+        if self.status.state != UserTaskStatus.FAILED:
+            self.status.fail({'raw_error_msg': str(exception)})
+        return
+
+
+def _create_broken_link_json(course_block, course_key, context, status=None):
+    """
+    Generates a json file for broken links, or returns None if there was an error.
+    Updates the context with any error information if applicable.
+    Note that because the celery queue does not have credentials, some broken links will
+    need to be checked client side.
+
+    File format:
+    [
+        [<blockId>, <broken link>],
+        ...,
+    ]
+    """
+    name = course_block.url_name
+    links_file = NamedTemporaryFile(prefix=name + '.', suffix='.json')
+
+    try:
+        if status:
```

Comment:
You're using `status` as a sort of state in a state machine in this solution. It's worth considering extracting code that changes status into its own method (say, `update_status()`. Doing so makes it very clear to the reader what's happening, as the one line call contains just two things: a method name that tells you you're updating status, and the new status, and nothing else. There's also less code duplication as you change status in various parts of your code. But, perhaps most importantly, having a method that changes status is really handy when using the debugger, as putting a breakpoint at the entry to that method allows you to easily break on reaching a desired application state.